Fix fluentdconfig and syslogngconfig deletion

OverOrion opened the pull request:
This PR aims to fix two problems:
- deleting a fluentdconfig/syslogngconfig resource should also delete the child resources (solved by changing the ownerreferences)
- deleting a logging resource should not be possible until a fluentdconfig/syslogngconfig resource exists. we don't delete the fluentdconfig/syslogngconfig forcefully because those are created by the user, but we can protect against this by using a finalizer. we send an event to the logging object for the user to be able to see other than looking at the operator logs

Fixes #1667.

Signed-off-by: Szilard Parrag <szilard.parrag@axoflow.com>

pepov commented on controllers/logging/logging_controller.go:
```diff
@@ -128,6 +130,18 @@ func (r *LoggingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return reconcile.Result{}, errors.WrapIfWithDetails(err, "failed to get logging resources", "logging", logging)
 	}
 
+	if ExternalFluentdRef := slices.Index(logging.Finalizers, logging.Status.FluentdConfigName); ExternalFluentdRef != -1 {
```
I'm missing the deletiontimestamp check from here

pepov replied:
also we should put up the finalizer somewhere here I beleive to keep it in one place

pepov commented on controllers/logging/logging_controller.go:
```diff
@@ -128,6 +130,18 @@ func (r *LoggingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+
+		fluentdConfigRefCleaned := r.cleanupFluentdConfigReference(ctx, log, &logging, fluentdConfigName)
+
+		if !fluentdConfigRefCleaned && logging.DeletionTimestamp != nil {
```
why do we need the deletiontimestamp check here?

pepov commented on pkg/resources/model/reconciler.go:
```diff
@@ -235,6 +235,7 @@ func NewValidationReconciler(
 			}
 			logger.Info("found detached fluentd aggregator, making association", "name", resources.Fluentd.Configuration.Name)
 			resources.Logging.Status.FluentdConfigName = resources.Fluentd.Configuration.Name
+			resources.Logging.Finalizers = append(resources.Logging.Finalizers, resources.Fluentd.Configuration.Name)
```
I'm not sure this is the right place to do this. finalizer operations should be done at a single place

OverOrion commented:
I have reworked the PR according to pepov's notes and using kubebuilder's docs: https://book.kubebuilder.io/reference/using-finalizers.

I also added a secondary fix which occured when logging switched from `fluentdConfig` to inline spec.

pepov commented on controllers/logging/logging_controller.go:
```diff
@@ -274,10 +285,53 @@ func (r *LoggingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+		// Marked for deletion, check for fluentdConfig
+		fluentdConfigName := types.NamespacedName{Namespace: logging.Spec.ControlNamespace, Name: logging.Status.FluentdConfigName}
+		if controllerutil.ContainsFinalizer(logging, fluentdConfigFinalizer) && r.isExistingFluentdConfig(ctx, fluentdConfigName) {
+			return reconcile.Result{}, errors.NewWithDetails("failed to delete logging resources, delete fluentdConfig first", "fluentdConfig", logging.Status.FluentdConfigName)
```
I'm thinking about this hard because the user won't have any feedback about this. How about one of the following?
- we could actually remove the fluentdconfig resource here
- we could update the logging status so that the user see what's happening

or maybe it would be enough to use a logging ownerref on the fluentdconfig?

OverOrion replied:
We can not update the status once an object is marked for deletion: https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/#how-finalizers-work (see the note section)

Removing the fluentdconfig might be unexpected (admin deletes logging, tenant loses it's fluentdconfig object)

pepov commented on controllers/logging/logging_controller.go:
```diff
@@ -274,10 +285,53 @@ func (r *LoggingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+			return reconcile.Result{}, errors.NewWithDetails("failed to delete logging resources, delete fluentdConfig first", "fluentdConfig", logging.Status.FluentdConfigName)
+		}
+		controllerutil.RemoveFinalizer(logging, fluentdConfigFinalizer)
+		if err := r.Update(ctx, logging); err != nil {
```
what happens to the error here?

orymate commented on controllers/logging/logging_controller.go:
```diff
@@ -332,6 +376,14 @@ func (r *LoggingReconciler) isExistingFluentdConfig(ctx context.Context, fluentd
 }
 
+func (r *LoggingReconciler) isExistingSyslogNGConfig(ctx context.Context, syslogNGConfigref types.NamespacedName) bool {
+	var fluentdConfig loggingv1beta1.SyslogNGConfig
+	if err := r.Client.Get(ctx, syslogNGConfigref, &fluentdConfig); err != nil && apierrors.IsNotFound(err) {
```
```suggestion
	var config loggingv1beta1.SyslogNGConfig
	if err := r.Client.Get(ctx, syslogNGConfigref, &config); err != nil && apierrors.IsNotFound(err) {
```

pepov commented on controllers/logging/logging_controller.go:
```diff
@@ -302,15 +309,21 @@ func (r *LoggingReconciler) checkFluentdConfigFinalizer(ctx context.Context, log
 }
 
+func (r *LoggingReconciler) cleanupFluentdConfigReference(ctx context.Context, logging *loggingv1beta1.Logging, log logr.Logger) (ctrl.Result, error) {
+	log.Info("cleaned up fluentdConfigRef", "name", logging.Status.FluentdConfigName)
```
reworded
